src/tools/ceph_immutable_object_cache/SimplePolicy.hpp

```diff
@@ -0,0 +1,163 @@
+
+
+class SimplePolicy : public Policy {
+public:
+  SimplePolicy(uint64_t block_num, float watermark)
+    : m_watermark(watermark), m_entry_count(block_num),
+      m_cache_map_lock("rbd::cache::SimplePolicy::m_cache_map_lock"),
+      m_free_list_lock("rbd::cache::SimplePolicy::m_free_list_lock")
+  {
+
+    for(uint64_t i = 0; i < m_entry_count; i++) {
+      m_free_list.push_back(new Entry());
+    }
+
+  }
+
+  ~SimplePolicy() {
+    for(uint64_t i = 0; i < m_entry_count; i++) {
+      Entry* entry = reinterpret_cast<Entry*>(m_free_list.front());
+      delete entry;
+      m_free_list.pop_front();
+    }
+  }
+
+  CACHESTATUS lookup_object(std::string cache_file_name) {
+
+    //TODO(): check race condition
+    RWLock::WLocker wlocker(m_cache_map_lock);
+
+    auto entry_it = m_cache_map.find(cache_file_name);
+    if(entry_it == m_cache_map.end()) {
+      Mutex::Locker locker(m_free_list_lock);
+      Entry* entry = reinterpret_cast<Entry*>(m_free_list.front());
+      assert(entry != nullptr);
+      m_free_list.pop_front();
+      entry->status = OBJ_CACHE_PROMOTING;
+
+      m_cache_map[cache_file_name] = entry;
+
+      return OBJ_CACHE_NONE;
+    }
+
+    Entry* entry = entry_it->second;
+
+    if(entry->status == OBJ_CACHE_PROMOTED) {
+      // touch it
+      m_promoted_lru.lru_touch(entry);
+    }
+
+    return entry->status;
+  }
+
+  int evict_object(std::string& out_cache_file_name) {
+    RWLock::WLocker locker(m_cache_map_lock);
+
+    return 1;
+  }
+
+  // TODO(): simplify the logic
+  void update_status(std::string file_name, CACHESTATUS new_status) {
+    RWLock::WLocker locker(m_cache_map_lock);
+
+    Entry* entry;
+    auto entry_it = m_cache_map.find(file_name);
+
+    // just check.
+    if(new_status == OBJ_CACHE_PROMOTING) {
+      assert(entry_it == m_cache_map.end());
+    }
+
+    assert(entry_it != m_cache_map.end());
+
+    entry = entry_it->second;
+
+    // promoting is done, so update it.
+    if(entry->status == OBJ_CACHE_PROMOTING && new_status== OBJ_CACHE_PROMOTED) {
+      m_promoted_lru.lru_insert_top(entry);
+      entry->status = new_status;
+      return;
+    }
+
+    assert(0);
+  }
+
+  // get entry status
+  CACHESTATUS get_status(std::string file_name) {
+    RWLock::RLocker locker(m_cache_map_lock);
+    auto entry_it = m_cache_map.find(file_name);
+    if(entry_it == m_cache_map.end()) {
+      return OBJ_CACHE_NONE;
+    }
+
+    return entry_it->second->status;
+  }
+
+  void get_evict_list(std::list<std::string>* obj_list) {
+    RWLock::WLocker locker(m_cache_map_lock);
+    // check free ratio, pop entries from LRU
+    if (m_free_list.size() / m_entry_count < m_watermark) {
+      int evict_num = 10; //TODO(): make this configurable
+      for(int i = 0; i < evict_num; i++) {
+        Entry* entry = reinterpret_cast<Entry*>(m_promoted_lru.lru_expire());
+        if (entry == nullptr) {
+	  continue;
+        }
+        std::string file_name = entry->cache_file_name;
```

Comment:
`entry->cache_file_name` was never specified, so `file_name` would always return an empty string. A quick and easy fix would be to assign `entry->cache_file_name` when putting `entry` into `m_cache_map`.

Reply:
@usefulalgorithm  i don't understands what's your meaning here? why entry->cache_file_name was never specified ? what's situation will occur this case ?

Reply:
It will return an empty string. Object `entry` has a member called `cache_file_name` but it was never assigned. 

I fixed this by adding this line in the function `CACHESTATUS lookup_object(...)`:

```
...
   m_cache_map[cache_file_name] = entry;
+   entry->cache_file_name = cache_file_name;
...
```

Not sure if this is an appropriate fix though.